num.nix

```diff
@@ -153,4 +209,446 @@ rec {
+    else
+      let
+        go = xs: n:
+          if n < radix
+          then list.cons n xs
+          else
+            let qr = quotRem n radix;
+            in go (list.cons (qr._1) xs) qr._0;
+      in go [] x;
+
+  /* fromBaseDigits :: int -> [int] -> int
+
+     Convert a list of digits in the given base into an int. The most
+     significant digit is the first element of the list.
+  */
+  fromBaseDigits = radix: list.foldl' (acc: n: acc * radix + n) 0;
+
+  /* toHexString :: int -> string
+  */
+  toHexString = x:
+    let toHexDigit = string.index "0123456789ABCDEF";
+    in string.concatMap toHexDigit (toBaseDigits 16 x);
+
+  /* gcd :: int -> int -> int
+
+     Computes the greatest common divisor of two integers.
+  */
+  gcd = x: y:
+    let gcd' = a: b: if b == 0 then a else gcd' b (rem a b);
+    in gcd' (abs x) (abs y);
+
+  /* lcm :: int -> int -> int
+
+     Computes the least common multiple of two integers.
+  */
+  lcm = x: y:
+    if x == 0 || y == 0
+    then 0
+    else abs (quot x (gcd x y) * y);
+
+  /* clamp :: num -> num -> num
+
+     Clamp the value of the third argument to be between the first two
+     arguments, so that the result is lower-bounded by the first argument and
+     upper-bounded by the second.
+  */
+  clamp = lo: hi: x: min (max lo x) hi;
+
+  /* maxInt :: int
+
+     The highest representable positive integer.
+  */
+  maxInt = 9223372036854775807; # 2^63 - 1
+
+  /* minInt :: int
+
+     The lowest representable negative integer.
+  */
+  minInt = maxInt + 1; # relies on 2's complement representation
+
+  bits =
+    let
+      # table of (bit n) for 0 <= n <= (sizeof(int) - 1)
+      powtab = [
+        1 2 4 8 16 32 64 128 256 512 1024 2048 4096 8192 16384 32768 65536
+        131072 262144 524288 1048576 2097152 4194304 8388608 16777216 33554432
+        67108864 134217728 268435456 536870912 1073741824 2147483648 4294967296
+        8589934592 17179869184 34359738368 68719476736 137438953472 274877906944
+        549755813888 1099511627776 2199023255552 4398046511104 8796093022208
+        17592186044416 35184372088832 70368744177664 140737488355328
+        281474976710656 562949953421312 1125899906842624 2251799813685248
+        4503599627370496 9007199254740992 18014398509481984 36028797018963968
+        72057594037927936 144115188075855872 288230376151711744
+        576460752303423488 1152921504606846976 2305843009213693952
+        4611686018427387904 (9223372036854775807 + 1)
+      ];
+      # De Bruijn multiplication lookup table, used for bitScanReverse
+      debruijn = [
+        0 47 1 56 48 27 2 60
+        57 49 41 37 28 16 3 61
+        54 58 35 52 50 42 21 44
+        38 32 29 23 17 11 4 62
+        46 55 26 59 40 36 15 53
+        34 51 20 43 31 22 10 45
+        25 39 14 33 19 30 9 24
+        13 18 8 12 7 6 5 63
+      ];
+    in rec {
+      /* bitSize :: int
+
+         The number of bits used to represent an integer in Nix.
+      */
+      bitSize = 64;
```

Comment:
it's possible that we can have a function `bitSize :: number -> maybe int` that uses type discrimination to return the number of bits. wait, we can't do that, nevermind, because we can't tell the difference between eg a word8 and word16. keeping this comment for posterity.

Reply:
some of the bitwise operations do rely on the [representation](https://github.com/NixOS/nix/blob/master/src/libexpr/value.hh#L61) more than I would like, like some of them require hardcoding the bit size for efficiency's sake, but I think this is mostly unavoidable given nix's limited builtins and such.